### ISSUE TITLE
[.WebGL-0x7f89f98b0000]GL ERROR :GL_INVALID_OPERATION : glDrawElementsInstancedANGLE: attempt to access out of range vertices in attribute 2

### ISSUE THREAD

**Issue by jamesleesaunders** [x3dom/x3dom#976]
Possible regression from 1.7.2? 
I have an interactive scatter plot generated with X3DOM and D3 which also uses click and hover events.
Expected Behaviour: On hovering over bubbles a label is shown. On clicking on a bubble a crosshair is shows.

Under X3DOM v1.8.0 it no longer behaves as previously in v1.7.2. 
When clicking on a bubble the background briefly flickers black and a JavaScript console error is generated:
```[.WebGL-0x7f89f98b0000]GL ERROR :GL_INVALID_OPERATION : glDrawElementsInstancedANGLE: attempt to access out of range vertices in attribute 2```
This only happens on the first click. From then on it behaves fine.

Chrome Desktop on OSX.

With the current v1.0.8 (dev)
https://raw.githack.com/jamesleesaunders/d3-x3d/scatter-bug/examples/X3DOM/chart/ScatterPlot.html

With v1.7.2:
https://raw.githack.com/jamesleesaunders/d3-x3d/scatter-bug/examples/X3DOM/chart/ScatterPlot_1_7_2.html

**Comment by andreasplesch** [comment]
Confirmed on Chrome for linux.
It will be probably necessary to define an as minimal as possible example. For example, does a x3dom test or x3dom example generate the flickering ?

**Comment by jamesleesaunders** [comment]
Confirmed also an issue on Chrome+Windows & Firefox+OSX. 
But appears not to be an issue on Safari+OSX. Nor is it an issue on my older version of Firefox+Windows.

I will try and concoct a simpler example to see if I can illuminate any possible bugs in my d3-x3d library.

**Comment by andreasplesch** [comment]
OSX and  older firefox may indicate that it is webgl2 related.

**Comment by jamesleesaunders** [comment]
I have put together a couple of more cut down examples and think I may have possibly narrowed it down to the presence of a 'Background' element.

**Test 1:**
  https://raw.githack.com/jamesleesaunders/d3-x3d/scatter-bug/examples/X3DOM/chart/Test1.html
Click on the ball, a cylinder is attached and we still see the black background flickering effect & JS error.

**Test 2:**
  https://raw.githack.com/jamesleesaunders/d3-x3d/scatter-bug/examples/X3DOM/chart/Test2.html
Everything is the same except I have removed the '<Background groundColor="1 1 1" skyColor="1 1 1"></Background>' element. In this example we do not see the flickering and no error.

These tests still use D3 (rather than vanilla attachEventHandler) but it is the same effect.

**Comment by jamesleesaunders** [comment]
If any relevance...

Test 3:
https://raw.githack.com/jamesleesaunders/d3-x3d/scatter-bug/examples/X3DOM/chart/Test3.html
I had a go at appending different shapes to the scene instead of cylinder. We get the same flickering effect if we append a Box, however no flickering if we append a Sphere.

**Comment by jamesleesaunders** [comment]
To further simplify things:

**Test 4:**
https://codepen.io/jamesleesaunders/pen/qeJaRa
To eliminate any issue with X3DOM event / onclick mouse over issues I have instead added a separate button outside the canvas to append the cylinder. The same flickering is seen.

**Test 5:**
https://codepen.io/jamesleesaunders/pen/YmJGZQ
Finally to eliminate our previous gremlins introduced though the use of .append()/.attr() I have switched to use .html() - again we see the same flickering background

**Comment by tsturm** [comment]
Currently I can't reproduce the Issue on Windows + Chrome, Windows + Firefox, Windows + Edge.

@jamesleesaunders, @andreasplesch Can one of you post your output of ```x3dom.caps``` ?

**Comment by jamesleesaunders** [comment]
x3dom.caps 
```
{PLATFORM: "MacIntel", AGENT: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_5) Ap…L, like Gecko) Chrome/76.0.3809.100 Safari/537.36", RENDERMODE: "HARDWARE", DOMNodeInsertedEvent_perSubtree: true, MOBILE: false, …}
AGENT: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.100 Safari/537.36"
ALIASED_LINE_WIDTH_RANGE: Float32Array(2) [1, 1]
ALIASED_POINT_SIZE_RANGE: Float32Array(2) [1, 255.875]
ALPHA_BITS: 8
ANISOTROPIC: EXTTextureFilterAnisotropic {}
BACKEND: "webgl"
BLUE_BITS: 8
COLOR_BUFFER_FLOAT: WebGLColorBufferFloat {}
COMPRESSED_TEXTURE: WebGLCompressedTextureS3TC {}
COMPRESSED_TEXTURE_FORMATS: Uint32Array []
DEBUGRENDERINFO: WebGLDebugRendererInfo {}
DEPTH_BITS: 24
DEPTH_TEXTURE: null
DOMNodeInsertedEvent_perSubtree: true
DRAW_BUFFERS: WebGLDrawBuffers {}
EXTENSIONS: (28) ["ANGLE_instanced_arrays", "EXT_blend_minmax", "EXT_color_buffer_half_float", "EXT_disjoint_timer_query", "EXT_float_blend", "EXT_frag_depth", "EXT_shader_texture_lod", "EXT_texture_filter_anisotropic", "WEBKIT_EXT_texture_filter_anisotropic", "EXT_sRGB", "OES_element_index_uint", "OES_standard_derivatives", "OES_texture_float", "OES_texture_float_linear", "OES_texture_half_float", "OES_texture_half_float_linear", "OES_vertex_array_object", "WEBGL_color_buffer_float", "WEBGL_compressed_texture_s3tc", "WEBKIT_WEBGL_compressed_texture_s3tc", "WEBGL_compressed_texture_s3tc_srgb", "WEBGL_debug_renderer_info", "WEBGL_debug_shaders", "WEBGL_depth_texture", "WEBKIT_WEBGL_depth_texture", "WEBGL_draw_buffers", "WEBGL_lose_context", "WEBKIT_WEBGL_lose_context"]
FPL_TEXTURES: OESTextureFloatLinear {}
FP_TEXTURES: OESTextureFloat {}
GREEN_BITS: 8
HFPL_TEXTURES: OESTextureHalfFloatLinear {}
HFP_TEXTURES: false
INDEX_UINT: true
INSTANCED_ARRAYS: ANGLEInstancedArrays {}
MAX_ANISOTROPY: 16
MAX_COMBINED_TEXTURE_IMAGE_UNITS: 80
MAX_CUBE_MAP_TEXTURE_SIZE: 16384
MAX_RENDERBUFFER_SIZE: 16384
MAX_TEXTURE_IMAGE_UNITS: 16
MAX_TEXTURE_SIZE: 16384
MAX_VARYING_VECTORS: 15
MAX_VERTEX_ATTRIBS: 16
MAX_VERTEX_TEXTURE_IMAGE_UNITS: 16
MAX_VERTEX_UNIFORM_VECTORS: 1024
MAX_VIEWPORT_DIMS: Int32Array(2) [16384, 16384]
MOBILE: false
PLATFORM: "MacIntel"
RED_BITS: 8
RENDERER: "WebKit WebGL"
RENDERMODE: "HARDWARE"
SAMPLES: 4
SHADING_LANGUAGE_VERSION: "WebGL GLSL ES 1.0 (OpenGL ES GLSL ES 1.0 Chromium)"
STD_DERIVATIVES: OESStandardDerivatives {}
TEXTURE_LOD: false
UNMASKED_RENDERER_WEBGL: "Intel(R) Iris(TM) Graphics 6100"
UNMASKED_VENDOR_WEBGL: "Intel Inc."
VENDOR: "WebKit"
VERSION: "WebGL 1.0 (OpenGL ES 2.0 Chromium)"
WEBGL_VERSION: 1
}
```

**Comment by andreasplesch** [comment]
Using test case 5)
https://codepen.io/jamesleesaunders/pen/YmJGZQ

I see a white background first.

Then I click on the button and the background turns black. as long as the mouse does not enter the canvas. I think the background should stay white and this is what is seen as flickering in other examples.

Chrome on Windows 10.

**Comment by jamesleesaunders** [comment]
@tsturm That is annoying you have not managed to recreate the bug yet. 
Here are some screenshot gifs showing the issue:

The original 'flickering' as seen on the scatter plot:
![Aug-10-2019 19-54-52](https://user-images.githubusercontent.com/13055416/62825783-0f2d3980-bba9-11e9-9a17-1afcbb898fed.gif)

The cut-down version (test case 5 above) also showing glitches:
![Aug-10-2019 19-56-39](https://user-images.githubusercontent.com/13055416/62825779-05a3d180-bba9-11e9-8150-8437105de654.gif)

As @andreasplesch mentions, the canvas background briefly turns black until the move moves over the canvas at which point it flicks back to white.

**Comment by tsturm** [comment]
I will try to reproduce it on another device tomorrow.

**Comment by tsturm** [comment]
I found a missing eyeIdx binding.
Can you try if my last commit has fixed the issue?

**Comment by jamesleesaunders** [comment]
@tsturm Yes, it looks like the flickering background issue is now fixed and the error resolved.
Fixed by 86ed56b3baebe282bd184f1500f9f256388f1bf2.
Re-tested with Firefox + Chrome on OSX.

So I can at least try and understand what the issue, if it is simple to explain, was what is an eyeIdx binding? I may still be none the wiser!

Thanks again for all your hard work on x3dom!

**Comment by tsturm** [comment]
The eyeIdx is a seperate attribute which is used for the instance-based VR-Rendering.
Both eyes are rendered with a single draw call but the left & right eye get a different eyeIdx to decide which matrix we have to use and which part of the image can be discard.
